tutorial for chrome plugin (#82)
Knotx/knotx-website: src/render/tutorials/chrome-plugin/2_2.html.md

- fragments' structure (in fact, the diagram above is a screenshot from the tool)
- data flow in the graph
- additinal data for individual nodes, e.g. a response a node got from a service

The usage of the extension is explained throughout this tutorial. You can download it (TODO: link to the chrome store).

## Writing a simple fragment
As already mentioned, we won't be going into details about how to setup the project. For this tutorial, we have an HTML page like this:

![Raw Page](/img/blog/chrome-extension/raw-page.png)

As you can see, right now it's not even processed. There is a raw templating engine markup rendered. In code, it's just a plain HTML with two Knot.x snippets inside. One of them contains markup for User Details section, and the other for Payment Providers and Offers sections. It's worth noting that while the first snippet uses [Handlebars](https://handlebarsjs.com/), the second one uses [Pebble](https://pebbletemplates.io/) templating engine. It's possible for multiple markups to exist in the same HTML file, because snippets are processed independently from each other.

This is how the first snippet looks like in code:

> ```handlebars

Author: It's supported. Pebble is not though

> ```handlebars

Reviewer: Ok, somewhere below I saw usage of `pebble` syntax, so you should probably change it to something supported.

> ```handlebars

Author: I've defined `pebble` as alias for `html`

If we were to leave the configuration like this we would create a REST API. We want to render a page though, so we need another node in our task. When the first node succeeds (so user data is fetched), we want to call a node that renders a page. This is the updated code for the `user-task`:

```hocon
user-task {
  action = fetch-user-info
  on._success {
    action = te-hbs
  }
}
```

The `on` object contains all transitions (graph edges) we want to be able to handle. The format looks like this:

```hocon
on {
  transition1 = node1
  transition2 = node2
  // ...
}
```

In this case we define that after a node with action `fetch-user-info` Knot.x should call a node with action `te-hbs`. This action will render the page using handlebars. This is how it looks like:

```hocon
te-hbs {
  factory = knot
  config.address = knotx.knot.te.handlebars
}
```

Perhabs not what you were expecting. The `knot` factory allows for registering actions on the [Vert.x Event Bus](https://vertx.io/docs/vertx-core/java/#event_bus), thus providing more scalability. Here's how the actual handlebars configuration looks like:

```hocon

> address = knotx.knot.te.handlebars

Reviewer 2: Maybe we should provide some scalability options here:
```
modules {
  hbsTemplateEngine = "io.knotx.te.core.TemplateEngineKnot"
}
config.hbsTemplateEngine.options {
  instances = 4
  config {
    address = knotx.knot.te.handlebars
    engine {
      factory = handlebars
      config = {
        cacheSize = 1000
      }
    }
  }
}
```
?

Reviewer 2: Knot action (`te-hbs`) delegates the processing to "workers" that listens on a particular address (`knotx.knot.te.handlebars`). With that, we can fully utilise CPU resources (horizontal scaling) or use cluster mode to deploy modules vertically (on different machines) not affecting graph simplicity.

engine {
  factory = handlebars
  config = {
    cacheSize = 1000
  }
}
```

The entire configuration for the first snippet:

```hocon
// tasks.conf

user-task {
  action = fetch-user-info
  on._success {
    action = te-hbs
  }
}

// actions.conf

fetch-user-info {
  factory = http
  config.endpointOptions {
    path = /user
    host = webapi
    port = 8080
  }
}
te-hbs {
  factory = knot
  config.address = knotx.knot.te.handlebars
}
```

Now the User Details section of the first snippet should render like this:

![Rendered User Details](/img/blog/chrome-extension/filled-user-details.png)

Here's when the chrome extension comes in. If we have [Fragment HTML Body Writer](https://github.com/Knotx/knotx-fragments/tree/master/handler/consumer/html) configured 
adequately, the extension can read data about fragments from the HTML page. Assuming the extension is installed, let's open the inspector and go to the "Knot.x" tab. We can see a list of fragments on the page:

![Rendered User Details](/img/blog/chrome-extension/list-of-fragments.png)

> There are two fragments because we have two snippets in HTML. We haven't defined a task for the second one thus it displayes a grey square next to it. It means it wasn't processed. Ignore it for now and focus on the first one. It's green so it ended with a success. There's also a randomly generated ID, a type ("snippet" is the only possible fragment type in an HTML page), and the time it took to process the fragment. If you click on the arrow pointing down next to the ID, you should see a list of top level HTML elements in the snippet. In this case it's just one div. Hover over it to highlight the div on the page.

Reviewer 2: > "snippet" is the only possible fragment type in an HTML page

this is not true, we can define more types such as: `hbs-snippet` or `pebble-snippet`

Then in Fragment HTML Body Writer we configure the [fragmentType](https://github.com/Knotx/knotx-fragments/tree/master/handler/consumer/html#how-to-configure) option.

Fragment details is the most interesting and useful part of the extension. Click on the fragment on the list to view its details. It should look like this:

![Rendered User Details](/img/blog/chrome-extension/first-snippet-graph.png)

You can drag and zoom using your mouse. The structure of the displayed graph matches the structure we defined in the `user-task` task. Click on the `fetch-user-info` node to see the details of its execution. We can, for example, see what request it sent and exactly what response it got in the `response` property. We can also view node's configuration in the `operation` property.

> ![Node Details](/img/blog/chrome-extension/node-details.png)

Reviewer 2: You can expand the request too.

    path = /payments
    domain = webapi
    port = 8080
  }
}
fetch-offers {
  factory = http
  config.endpointOptions {
    path = /offers
    domain = webapi
    port = 8080
  }
}
te-pebble {
  factory = knot
  config.address = knotx.knot.te.pebble
}
```

Nothing we haven't seen already. When we load the page now we should see that the second snippet is filled with data. In the Knot.x extension we still see two fragments but this time they both have green boxes next to them - both were processed and ended with success. Click the second fragment on the list to view its details:

![Naive Approach](/img/blog/chrome-extension/naive-second-snippet.png)

It works, but it's not the best approach. Right now the `fetch-offers` node waits for the `fetch-payment-providers` to succeed. There is no reason for that (remmeber, we assumed that those pieces of information can be retrieve intependently). There would also be a problem if `fetch-payment-providers` failed for some reason. Then `fetch-offers` would not execute, even if it would be able to retrieve data and show them to the user. It'd be more efficient and safer if those two nodes could execute in parallel.

> There are two types of nodes. Everything you've seen up to this point were so called "single nodes". They just execute an action. The second type is a composite node. This kind of node contains an array of other nodes. It executes them in parallel and waits for all of them to complete. If all of them succeed then the composite succeeds. Otherwise the composite fails. This is how we can reorder our nodes to make use of a composite:

Reviewer 2: > This kind of node contains an array of other nodes.

This kind of node contains an array of other root nodes (in fact, those are subgraphs).

```hocon
payments-task {
  action = fetch-user-info

  on._success {
    actions = [
      {
        action = fetch-payment-providers
      },
      {
        action = fetch-offers
      }
    ]

    on._success {
      action = te-pebble
    }
  }
}
```

After `fetch-user-info` succeeds the composite is started to be processed. It calls `fetch-payment-providers` and `fetch-offers` in parallel. When both of them succeeds the pebble node starts to be processed. We don't need to change any actions, because we've only modified the structure of the graph and not what we do with data.

When we load the page now nothing seems to have changed. However, when we open the chrome extension we can see that the second fragment looks like this:

![Composite](/img/blog/chrome-extension/composite.png)

The white circle represents the beginning of a composite. The second circle is where the composite ended. All subtasks (nodes in the composite) ended with success thus the composite did too. That's why the second circle is green.

> This is a good moment to introduce the perfomance tab. When you open the tab called "Perfomance View" you'll see a timeline of node execution:

Reviewer: ```suggestion
This is a good moment to introduce the performance tab. When you open the tab called "Performance View" you'll see a timeline of node execution:
```

![Timeline](/img/blog/chrome-extension/timeline.png)

> By looking at this it's clear that `fetch-payment-providers` and `fetch-offers` executed simultanously and that the next node (`te-pebble`) executed only after both of them finished.

Reviewer: ```suggestion
By looking at this it's clear that `fetch-payment-providers` and `fetch-offers` executed simultaneously and that the next node (`te-pebble`) executed only after both of them finished.

```

This is far better from what we had before. There is still room for improvement though. Right now we don't handle any errors. In a real application errors are unavoidable. Let's say that the service that we call to retrieve offers has a long response time. We can simulate it by using a mock `/timeout`. It responds with the same data as `/offers` but it takes three seconds to do so. Swap the endpoint in the `fetch-offers` action:

```hocon
fetch-offers {
  factory = http
  config.endpointOptions {
    path = /timeout
    domain = webapi
    port = 8080
  }
}
```

> When we load the page now it should take about three seconds. If we open the Knot.x extension we can immediatly spot on the fragment list that the second fragment executed for a long time. When we open the fragments performance view we can clearly see that the problem lays in the composite and specificaly in the `fetch-offers` node.

Reviewer: ```suggestion
When we load the page now it should take about three seconds. If we open the Knot.x extension we can immediately spot on the fragment list that the second fragment executed for a long time. When we open the fragments performance view we can see that the problem lays in the composite and specifically in the `fetch-offers` node.
```

> What if such a long time is unacceptable? One approach would be to define timeout in the `fetch-offers` action. However, we can do better. Knot.x provides us with a built-in implementation of a [Circuit Breaker Pattern](https://en.wikipedia.org/wiki/Circuit_breaker_design_pattern). It does so by a [Circuit Breaker Factory](TODO-find-a-link-to-documentation) that acts as a wrapper. Let's see how it can be configured:

Reviewer: ```suggestion
What if such a long time is unacceptable? One approach would be to define a timeout in the `fetch-offers` action. However, we can do better. Knot.x provides us with a built-in implementation of a [Circuit Breaker Pattern](https://en.wikipedia.org/wiki/Circuit_breaker_design_pattern). It does so by a [Circuit Breaker Factory](TODO-find-a-link-to-documentation) that acts as a wrapper. Let's see how it can be configured:
```

```hocon
fetch-offers-cb {
  factory = cb
  doAction = fetch-offers

  config.circuitBreakerOptions {
    maxFailures = 3
    timeout = 500
    resetTimeout = 20000
  }
}
```

The circuit breaker calls the `fetch-offers` action inside of itself (`doAction` property). If the action failed then the circuit breaker assumes something is wrong and won't actually call the action next time it's invoked. It'll just assume failure instead.

We defined the following properties:
- `maxFailures` - number of failures it takes for circuit breaker to assume the action is not working
- `timeout` - the circuit breaker will assume failure after this time
- `resetTimeout` - how long the circuit breaker should wait until it'll allow the action to execute again

Now we need to adjust the task. This is how it should look like now:

```
payments-task {
  action = fetch-user-info

  on._success {
    actions = [
      {
        action = fetch-payment-providers
      },
      {
        action = fetch-offers-cb
      }
    ]

    on {
      _success {
        action = te-pebble
      }
      _error {
        action = te-pebble
      }
    }
  }
}
```

> Firstly, we call `fetch-offers-cb` instead of `fetch-offers` in the composite. That's because the `fetch-offers` action will be called inside of `fetch-offers-cb`. Also, since we expect the task to fail, the composite should handle the `_error` transition. It will be called if any subtask ends with anythink different then success. In this case we still want to render the page so we call the `te-pebble` action. 

Reviewer: ```suggestion
First, we call `fetch-offers-cb` instead of `fetch-offers` in the composite. That's because the `fetch-offers` action will be called inside of `fetch-offers-cb`. Also, since we expect the task to fail, the composite should handle the `_error` transition. It will be called if any subtask ends with anything different than success. In this case, we still want to render the page so we call the `te-pebble` action. 
```

> When we reload the page we can see that the Offers section is empty. However the Payment Providers section should still be filled with data. The tool will show this graph:

Reviewer: ```suggestion
When we reload the page we can see that the Offers section is empty. However, the Payment Providers section should still be filled with data. The tool will show this graph:
```

![Missing Node](/img/blog/chrome-extension/missing-node.png)

There's a lot of going on here. The `fetch-offers-cb` node is blue. That's because it returned a `_fallback` transition. It's a custom transition. Transitions are nothing more than strings. `_success` and `_error` are standard and they display respectively as green and red. Anything else is a custom transition and is displayed as blue. If we click on the `fetch-offers-cb` node we can see why it returned fallback: `Exception: operation timeout`.

![Circuit Breaker Details](/img/blog/chrome-extension/cb-details.png)

> The yellow node with an exclamation mark is a so called "missing node". It doesn't actually exist during processing. It's just displayed in the tool as a way to catch the developer's attention. It means that a transition was triggered (in this case `_fallback`) but there was no node to handle it. Because of that, the `_fallback` state was returned to the composite and thus the composite ended with an error (composites can only end with either `_success` or `_error`).

Reviewer: ```suggestion
The yellow node with an exclamation mark is a so-called "missing node". It doesn't exist during processing. It's just displayed in the tool as a way to catch the developer's attention. It means that a transition was triggered (in this case `_fallback`) but there was no node to handle it. Because of that, the `_fallback` state was returned to the composite and thus the composite ended with an error (composites can only end with `_success` or `_error`).
```

Also, note the dashed transitions. They represent a transition that could potentially be triggered, but in this case wasn't.

Finally, let's implement a node to handle the fallback case. The task should now look like this:

```hocon
payments-task {
  action = fetch-user-info

  onTransitions._success {
    actions = [
      {
        action = fetch-payment-providers
      },
      {
        action = fetch-offers-cb
        
        on._fallback {
          action = fetch-offers-fallback
        }
      }
    ]

    onTransitions {
      _success {
        action = te-pebble
      }
      _error {
        action = te-pebble
      }
    }
  }
}
```

We've added an action called `fetch-offers-fallback` that will be triggered when the circuit breaker returns a `_fallback` transition. This is a transition for the composite's subtask and not for the composite.

> The `fetch-offers-fallback` action could retrieve data from cache or from another service. In our case it'll just call the `/offers` endpoint:

Reviewer: ```suggestion
The `fetch-offers-fallback` action could retrieve data from cache or another service. In our case it'll just call the `/offers` endpoint:
```

    domain = webapi
    port = 8080
  }
}
```

When we refresh the page we still won't see any offers displayed on the page. We'll get to this in a minute. The graph should look like this:

![Fallback](/img/blog/chrome-extension/with-fallback.png)

As we can see, even though the circuit breaker ended with fallback, the composite as a whole ended with success because of the `fetch-offers-fallback` node.

The only problem left is that we can't see any offers displayed, even though they were retrieved (you can check they were by viewing details of the fallback node). Let's take a look at the HTML snippet:

```html
<knotx:snippet data-knotx-task="payments-task">
  <div>
    <h2>Payment providers</h2>
    <ul>
      {% for provider in data['fetch-payment-providers']._result %}
      <li>
        <p>Payment provider: {{provider.name}}</p>
        <p>Payment URL: {{provider.paymentUrl}}</p>
      </li>
      {% endfor %}
    </ul>

    <h2>Offers</h2>
    <ul>
      {% for offer in data['fetch-offers']._result %}
      <li>
        <p>Offer: {{offer.method}}</p>
        <p>Price: {{offer.price}}</p>
      </li>
      {% endfor %}
    </ul>
  </div>
</knotx:snippet>
```

> Pebble retrieves offers data from a key called `fetch-offers`. By default, every action saves its data under its name. The problem is that in case of a timeout the action that retrieves data is called `fetch-offers-fallback` and therefore that's where the data is stored. One way to resolve this would be to use an action that rewrites data from one key to another. For the sake of this tutorial we'll just create additional markup for fallback messages. The Offers section of the snippet will look like this:

Reviewer: ```suggestion
Pebble retrieves offers data from a key called `fetch-offers`. By default, every action saves data under its name. The problem is that in case of a timeout the action that retrieves data is called `fetch-offers-fallback` and therefore that's where the data is stored. One way to resolve this would be to use an action that rewrites data from one key to another. For the sake of this tutorial, we'll just create additional markup for fallback messages. The Offers section of the snippet will look like this:
```

```html
<h2>Offers</h2>
<ul>
  {% for offer in data['fetch-offers']._result %}
  <li>
    <p>Offer: {{offer.method}}</p>
    <p>Price: {{offer.price}}</p>
  </li>
  {% endfor %}
  {% for offer in data['fetch-offers-fallback']._result %}
  <li>
    <p>Fallback offer: {{offer.method}}</p>
    <p>Fallback price: {{offer.price}}</p>
  </li>
  {% endfor %}
</ul>
```

If we render the page now, we'll be able to see the fallback offers:

![Fallback Offers](/img/blog/chrome-extension/fallback-offers.png)

## Summary

> As you can see Knot.x Fragments allows us to easily define data flow when integrating with 3rd party services. In this example we rendered a page, however this mechanism can be used to build restful APIs as well. The Knot.x extension can be used during many processes of building an application. It can be used by developers for debugging purposes, new team members to familiarize themselves with the application's architecture, quality assurance/testers to detect flaws and identify root causes of bugs (is it a frontend, backend or integration problem?) or architects and business people to visualize the data flow and buisiness cases.

Reviewer: ```suggestion
As you can see Knot.x Fragments allows us to easily define data flow when integrating with 3rd party services. In this example, we rendered a page, but note this mechanism can be used to build restful APIs as well. The Knot.x extension can be used during many processes of building an application. It can be used by developers for debugging purposes, new team members to familiarize themselves with the application's architecture, quality assurance/testers to detect flaws and identify root causes of bugs (is it a frontend, backend or integration problem?) or architects and business people to visualize the data flow and business cases.
```
